crates/bevy_reflect/src/type_name.rs

```diff
@@ -0,0 +1,110 @@
+use crate::utility::GenericTypeNameCell;
+
+/// Provides the name of the type as a string.
+///
+/// This is a stable alternative to [`std::any::type_name`] whose output isn't guarenteed
+/// and may change between versions of the compiler.
+///
+/// This trait may be derived via [`#[derive(TypeName)]`][bevy_reflect_derive::TypeName].
+///
+/// ## Manual implementation
+///
+/// If you need to manually implement [`TypeName`], it's recommended to follow
+/// the example below (unless specifying a custom name).
+///
+/// ```ignore
+/// bevy_reflect::TypeName;
+///
+/// struct MyType;
+///
+/// impl TypeName for MyType{
+///     fn name() -> &'static str {
+///         concat!(module_path!(), "::", "MyType")
+///     }
+/// }
+/// ```
+///
+/// If your type is generic you must use
+/// [`GenericTypeNameCell`][crate::utility::GenericTypeNameCell].
+///
+/// ```ignore
+/// bevy_reflect::{TypeName, utility::GenericTypeNameCell};
+///
+/// struct MyType<T>(T);
+///
+/// impl<T: TypeName> TypeName for MyType<T> {
+///     fn name() -> &'static str {
+///         static CELL: GenericTypeNameCell = GenericTypeNameCell::new();
+///         CELL.get_or_insert::<Self, _>(|| {
+///             format!(concat!(module_path!(), "::MyType<{}>"), T::name())
+///         })
+///     }
+/// }
+/// ```
+pub trait TypeName: 'static {
+    /// Returns the name of the type.
+    ///
+    /// This is a stable alternative to [`std::any::type_name`] whose output isn't guarenteed
+    /// and may change between versions of the compiler.
+    fn name() -> &'static str;
```

Comment:
I think some more structure here would benefit other parts of the engine (and reflection as a whole):

```rust
trait TypePath {
  /// The full type path
  /// "my_crate::my_mod::MyType<Generics>"
  fn type_path() -> &'static str;
  /// The short type name, without generics
  /// "MyType"
  fn short_type_name_base() -> &'static str;
  /// The short type name, with generics
  /// "MyType<Generics>"
  fn short_type_name() -> &'static str;
  /// The full type path, minus the actual type:
  /// "my_crate::my_mod"
  fn module_path() -> &'static str;
  /// The crate name
  /// "my_crate"
  fn crate() -> &'static str;
}
```

Ideally, these could all be predefined "slices" of the same static str (to avoid multiple string allocations). I don't think this is "easy", but this type of hackery feels possible, given that workarounds for const-slicing arrays and const concatenating static strs are a thing.

I also think we should _force_ some consistency here as part of the type name derive. (and consider manual impls invalid if they dont follow the rules). ex:
* A crate name is required and it must match the actual name of the crate
* Valid "rust module path syntax" must be used. Ex: `my_crate::my_module::MyType`.
* You are allowed to "shorten" crate+module paths, but you can't remove them entirely. Ex: using `MyType` as the full path is not allowed.

Reply:
One area that would benefit from this structure is #5723, which currently embeds the short type name in the reflect info, which feels redundant with this effort.

Reply:
Having a single static str and slicing it is possible.

In the example bellow, `PATH` and `IDENT` can be provided by the user and other consts are computed by the derive macro (in real case, the macro don't create those consts and use directly the value, it's to make the example clearer).

The only value that is tricky to compute is `LIB_NAME_LEN`. If the user provides a custom path, we got a literal string, so the macro can compute the `LIB_NAME_LEN` value. But if not provided, we use `module_path` and we have no way except by using `const_str`.
Actually, there is a unstable feature [`TokenStream::expend_expr`](https://doc.rust-lang.org/proc_macro/struct.TokenStream.html#method.expand_expr) (https://github.com/rust-lang/rust/issues/90765) that expend macro inside it, so we could get the result of `module_path` as a literal string. But since this feature is pretty recent, I don't think we can relied on it.

```rust
pub struct Foo<T>(T);

// The path is by default `module_path!()`
// If the user override it, it must start with the lib name
// and must be a valid rust path (e.g. `my_lib::my_mod`).
const PATH: &'static str = module_path!();

// By default the ident of the type.
// If the user override it, it must be a valid ident.
const IDENT: &'static str = "Foo";

// Values computed by the derive macro
const IDENT_POS: usize = PATH.len() + 2;
const GENERIC_POS: usize = IDENT_POS + IDENT.len();
const LIB_NAME_LEN: usize = const_str::split!(PATH, "::")[0].len();

impl<T: TypePath> TypePath for Foo<T> {
    fn type_path() -> &'static str {
        // if non generic, `concat!(PATH, "::", IDENT)` is used instead.
        let type_path = format!("{}::{}<{}>", PATH, IDENT, T::type_path());
        // In real case the value is stored in a static variable to avoid recomputation.
        Box::leak(Box::new(type_path))
    }

    fn short_type_name_base() -> &'static str {
        &Self::type_path()[IDENT_POS..GENERIC_POS]
    }

    fn short_type_name() -> &'static str {
        &Self::type_path()[IDENT_POS..]
    }

    fn module_path() -> &'static str {
        &Self::type_path()[..PATH.len()]
    }

    fn crate_name() -> &'static str {
        &Self::type_path()[..LIB_NAME_LEN]
    }
}
```